use low-level API for control writes

tlyu opened the pull request:
Use more of the low-level USBD API to handle control writes. This depends on a bugfix to the low-level firmware. Defer the validation of the control request setup header until the Data OUT stage has completed. This allows the recvControl to be called as expected from within the Arduino API setup handlers of the class-specific drivers. There is a fixed-size static buffer to receive the Data OUT payload, shared with the Data IN payload.

This is part of a series of changes that will eventually allow us to remove a large amount of local duplication of the vendor ISR code. That kind of local code duplication is likely to be risky and introduce race conditions, no matter how carefully it's implemented.

obra commented:
I’m not a huge fan of using a   USBCore function directly like this, since it’s different than how other cores do it. What would it take to fix the existing API to be safe so that it better matches other Arduino cores?On Feb 8, 2023, at 8:22 PM, Taylor Yu ***@***.***> wrote:﻿Breaking API change: recvControl renamed to setCtlOutDest. This is necessary because the API now writes to the buffer after the function returns, so it can't be passed an automatic buffer. Renaming the API forces callers to make sure that they're using a non-automatic buffer, to avoid memory corruption that could otherwise result.
This is part of a series of changes that will eventually allow us to remove a large amount of local duplication of the vendor ISR code. That kind of local code duplication is likely to be risky and introduce race conditions, no matter how carefully it's implemented.
This mutually depends on keyboardio/KeyboardioHID#96.



You can view, comment on, or merge this pull request online at:
  https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23

Commit Summary

  1effc77  use low-level API for control writes


File Changes  (4 files)

  
    M
    cores/arduino/CDCACM.cpp
    (2)
  
  
    M
    cores/arduino/USBCore.cpp
    (162)
  
  
    M
    cores/arduino/USBCore.h
    (9)
  
  
    M
    libraries/KeyboardioHID
    (2)
  


Patch Links:

  https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23.patch
  https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23.diff


—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

tlyu commented:
I tried. It's not really possible to preserve the same API signature while using the vendor API, without breaking existing code. I kept running into lockups and crashes because my KeyboardioHID would get out of sync with the GD32 core, and writing to an automatic buffer would scribble on the return address. I think that preserving the existing API is asking for latent bugs that difficult to troubleshoot.

Basically, existing code that calls `USB_RecvControl` expects to call it from within a function that is called from an ISR that handles the SETUP transaction, and also expects the call to `USB_RecvControl` to block until the OUT transaction providing the data has completed. That doesn't work with the low-level vendor API, which expects to be given a buffer that a later iteration of the ISR will fill with the OUT transaction contents.

_Maybe_ if we had a working async framework, I could `yield()` and make the existing API signature work. I also don't want to open the async framework can of worms right now.

tlyu commented:
If it's just a matter of not wanting the `USBCore()` reference to appear directly in user code, I could add a macro for `USB_SetCtlOutDest`.

obra commented:
Actually, in KeyboardioHID, we already have the infrastructure for doing
that in HID-Settings.h, so we don't have a whole much of ifdefs in the
codebase just to change the function name for gd32.

In terms of  renaming recvControl, what was your thinking about renaming to
something that's fairly different from the original if the function/use
case remains pretty much the same?

I

On Wed, Feb 8, 2023 at 8:52 PM Taylor Yu ***@***.***> wrote:

> If it's just a matter of not wanting the USBCore() reference to appear
> directly in user code, I could add a macro for USB_SetCtlOutDest.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23#issuecomment-1423631396>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2DV5IXNWXIUVHCTB33WWRZY7ANCNFSM6AAAAAAUWB5EBA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tlyu commented:
I would really rather have incompatible code obviously fail to compile (due to a missing symbol) than randomly crash hours after flashing. The function signature is identical, but the memory semantics are very much not.

obra commented:
Understood. But it's not at all obvious that recvControl and setCtlOutDest
are in any way related.



On Wed, Feb 8, 2023 at 9:27 PM Taylor Yu ***@***.***> wrote:

> I would really rather have incompatible code obviously fail to compile
> (due to a missing symbol) than randomly crash hours after flashing. The
> function signature is identical, but the memory semantics are very much not.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23#issuecomment-1423655169>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2ESXUEADBGKWOHA6T3WWR55BANCNFSM6AAAAAAUWB5EBA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tlyu commented:
Do you have a better name in mind?

obra commented:
I was thinking something of the form fooRecvControl or recvControlFoo “Safe” isn’t super descriptive. There’s got to be a better word. On Feb 8, 2023, at 9:34 PM, Taylor Yu ***@***.***> wrote:﻿
Do you have a better name in mind?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

tlyu commented:
So there might be a way to preserve the existing semantics of `recvControl`. It involves deferring calling the class-specific handler until after the Data OUT stage has completed. This means it's not possible to reject an invalid SETUP stage as early as possible during control writes. The class handler would be called in the OUT completion handler, and given the SETUP stage data. It could then call `recvControl` to retrieve the Data OUT stage data that's already been received and ACKed. The result of the handler determines whether the status stage completes successfully or not.

Unfortunately, this requires patching the vendor firmware to stop ignoring the result from the class-specific Data stage completion handlers. This might break existing code that previously had not been consistently returning valid values from these handlers.

tlyu commented:
I have a working prototype of this approach, and will clean it up soon.

> So there might be a way to preserve the existing semantics of `recvControl`. It involves deferring calling the class-specific handler until after the Data OUT stage has completed. This means it's not possible to reject an invalid SETUP stage as early as possible during control writes. The class handler would be called in the OUT completion handler, and given the SETUP stage data. It could then call `recvControl` to retrieve the Data OUT stage data that's already been received and ACKed. The result of the handler determines whether the status stage completes successfully or not.
> 
> Unfortunately, this requires patching the vendor firmware to stop ignoring the result from the class-specific Data stage completion handlers. This might break existing code that previously had not been consistently returning valid values from these handlers.

tlyu committed:
fix USBD firmware to check ctlx_out result

The low-level vendor USBD firmware didn't check the result of ctlx_out
in the class driver. This prevented an implementation strategy in the
Arduino core that requires deferring handling the validation of the
control transaction setup header until the Data OUT stage had
completed.

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu committed:
use low-level API for control writes

Use more of the low-level USBD API to handle control writes. This
depends on a bugfix to the low-level firmware. Defer the validation of
the control request setup header until the Data OUT stage has
completed. This allows the recvControl to be called as expected from
within the Arduino API setup handlers of the class-specific drivers.
There is a fixed-size static buffer to receive the Data OUT payload,
shared with the Data IN payload.

This is part of a series of changes that will eventually allow us to
remove a large amount of local duplication of the vendor ISR code.
That kind of local code duplication is likely to be risky and
introduce race conditions, no matter how carefully it's implemented.

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu commented:
Updated with new strategy. Also no longer depends on keyboardio/KeyboardioHID#96.

obra commented:
Awesome.

On Fri, Feb 10, 2023 at 6:58 PM Taylor Yu ***@***.***> wrote:

> Updated with new strategy. Also no longer depends on
> keyboardio/KeyboardioHID#96
> <https://github.com/keyboardio/KeyboardioHID/pull/96>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/23#issuecomment-1426585953>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2HKG5RJ5EMJRXT7DVLWW357FANCNFSM6AAAAAAUWB5EBA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
